utils/compress/idbcompress.cpp

```diff
@@ -180,19 +186,20 @@ int IDBCompressInterface::uncompressBlock(const char* in, const size_t inLen, un
     uint8_t storedMagic;
     utils::Hasher128 hasher;
 
-    outLen = 0;
 
     if (inLen < 1)
     {
+        outLen = 0;
```

Comment:
It seems to make sense to put the resetting of `outLen` at the beginning of the function and remove it from everywhere except line 227

Reply:
Thanks for point on this, I see that the function signature is different than function semantics, which cause misunderstanding. Basically LZ4 1. requires a size of uncompressed buffer (the buffer where the decompressed data will be written) and 2. does not have `getUncompressedSize`. Originally `uncomressBlock` wrote the size of uncompressed buffer which was received from `getUncompressedSize` to `outLen`. Basically we don't need to write the size of uncompressed buffer, because CS operates on chunks (`uncompressed buffer`) with fixed size. For initial abbreviated extent - initial chunk is about 1m (depends on row width) and regular chunk is about 4m, so we know the size of uncompressed buffer in any time for segment files. (Does not work for socket and joinpartitions). So, the `outLen` actually contains the size of output buffer, which we cannot set to 0 at the beginning. Anyway, I think it was my mistake to keep old function signature, I'll update it, and make `outLen` param as a value, not a reference.

Reply:
I'm a little worried about writeengine/bulk/we_colbufcompressed.cpp:766, where the outLen is saved for future use (?). Will everything be all right there?

Reply:
I think so, this function uncompress chunk based on the chunk index, to find the appropriate chunk index it uses `locateBlock`, if the chunk size is different against hardcoded (4m) it will locate the wrong chunk.
```

470 void CompressInterface::locateBlock(unsigned int block,
471                                     unsigned int& chunkIndex,
472                                     unsigned int& blockOffsetWithinChunk) const
473 {
474     const uint64_t BUFLEN  = UNCOMPRESSED_INBUF_LEN;
475 
476     uint64_t byteOffset    = (uint64_t)block * BLOCK_SIZE;
477     uint64_t chunk         = byteOffset / BUFLEN;
478     uint64_t blockInChunk  = (byteOffset % BUFLEN) / BLOCK_SIZE;
479 
480     chunkIndex             = chunk;
481     blockOffsetWithinChunk = blockInChunk;
482 }

```

Reply:
Anyway thanks for point on this, looking more detailed on the code, after the call of this function we have 
`228     // Expand the compression buffer size if working with an abbrev extent, and
229     // the bytes we are about to add will overflow the abbreviated extent.
230     if ((fToBeCompressedCapacity < CompressInterface::UNCOMPRESSED_INBUF_LEN) &&
231             ((fNumBytes + writeSize + fillUpWEmptiesWriteSize) > fToBeCompressedCapacity) )
232     {
233         std::ostringstream oss;
234         oss << "Expanding abbrev to-be-compressed buffer for: OID-" <<
235             fColInfo->curCol.dataFile.fid <<
236             "; DBRoot-"   << fColInfo->curCol.dataFile.fDbRoot    <<
237             "; part-"     << fColInfo->curCol.dataFile.fPartition <<
238             "; seg-"      << fColInfo->curCol.dataFile.fSegment;
239         fLog->logMsg( oss.str(), MSGLVL_INFO2 );
240         fToBeCompressedCapacity = CompressInterface::UNCOMPRESSED_INBUF_LEN;
241     }
`
Probably locateBlock works, because block == 0 for abbreviated extend. But currently not sure, could it brake something or not.

Reply:
Thanks for explaining!

Reply:
@drrtuy can you please comment on part of code which @mariadb-AlexeyAntipovsky found (thanks a lot for this!), does the new api will break something or not, because it seems to me not, but I'm not sure.
The new API does not modify the `outLen` for `uncompressBlock` because LZ4 does not have the function `getUncompressedSize`, and it seems it all should work fine, because the chunk size is hardcoded.

The code in which I'm not super sure are:

bulk load.
```
742         // Uncompress the chunk into our 4MB buffer
743         size_t outLen = CompressInterface::UNCOMPRESSED_INBUF_LEN;
744         int rc = fCompressor->uncompressBlock(
745                      compressedOutBuf,
746                      fChunkPtrs[chunkIndex].second,
747                      fToBeCompressedBuffer,
748                      outLen);

768         fToBeCompressedCapacity = outLen;
769 
for the function
nt ColumnBufferCompressed::initToBeCompressedBuffer
```
it saves the `outLen` and stores it to `fToBeCompressedCapacity`. 

Then it checks for abbreviated extent:

```
228     // Expand the compression buffer size if working with an abbrev extent, and
229     // the bytes we are about to add will overflow the abbreviated extent.
230     if ((fToBeCompressedCapacity < CompressInterface::UNCOMPRESSED_INBUF_LEN) &&
231             ((fNumBytes + writeSize + fillUpWEmptiesWriteSize) > fToBeCompressedCapacity) )
232     {
233         std::ostringstream oss;
234         oss << "Expanding abbrev to-be-compressed buffer for: OID-" <<
235             fColInfo->curCol.dataFile.fid <<
236             "; DBRoot-"   << fColInfo->curCol.dataFile.fDbRoot    <<
237             "; part-"     << fColInfo->curCol.dataFile.fPartition <<
238             "; seg-"      << fColInfo->curCol.dataFile.fSegment;
239         fLog->logMsg( oss.str(), MSGLVL_INFO2 );
240         fToBeCompressedCapacity = CompressInterface::UNCOMPRESSED_INBUF_LEN;
241     `}`
```

As far as I understood the chunk size could be different than 4m only for abbreviated extent and after first write it expands to the full extent.
Also this part of code relies on function `locateBlock`, to find `chunk index`, not sure how it possible to find appropriate chunk, if written chunk has len different than 4m using this function

```
470 void CompressInterface::locateBlock(unsigned int block,
471                                     unsigned int& chunkIndex,
472                                     unsigned int& blockOffsetWithinChunk) const
473 {
474     const uint64_t BUFLEN  = UNCOMPRESSED_INBUF_LEN;
475 
476     uint64_t byteOffset    = (uint64_t)block * BLOCK_SIZE;
477     uint64_t chunk         = byteOffset / BUFLEN;
478     uint64_t blockInChunk  = (byteOffset % BUFLEN) / BLOCK_SIZE;
479 
480     chunkIndex             = chunk;
481     blockOffsetWithinChunk = blockInChunk;
482 }

```

Tried different cases to trigger this part of code using `develop` branch and this branch
and chunk size is 4194304
Insert 1 row using cpimort into empty column, or into not empty column, and all work fine.

Could you please comment on this change?
Thanks!

Reply:
@mariadb-AlexeyAntipovsky @drrtuy found how to fix it, @mariadb-AlexeyAntipovsky thanks for review!

Reply:
I'm sorry for unresolving this but it seems that it would be better to store original `outLen` into a local variable, set `outLen` to 0 and remove zeroing every where. So the actual uncompress part (lines 222-229) will be something like
```
        auto rc = uncompress(&in[HEADER_SIZE], storedLen,
                             reinterpret_cast<char*>(out), &tmpOutLen);
        if (rc == ERR_OK)
        {
            outLen = tmpOutLen;
        }
        else
        {
            cerr << "uncompressBlock failed!" << endl;
            return ERR_DECOMPRESS;
        }
```

Reply:
Thanks, fixed.